src/python/pants/bin/daemon_pants_runner.py

```diff
@@ -110,31 +111,46 @@ def _nailgunned_stdio(self, sock):
     # Determine output tty capabilities from the environment.
     stdin_isatty, stdout_isatty, stderr_isatty = NailgunProtocol.isatty_from_env(self._env)
 
-    # Launch a thread to read stdin data from the socket (the only messages expected from the client
-    # for the remainder of the protocol), and threads to copy from stdout/stderr pipes onto the
-    # socket.
-    with NailgunStreamWriter.open_multi(
-           sock,
-           (ChunkType.STDOUT, ChunkType.STDERR),
-           None,
-           (stdout_isatty, stderr_isatty)
-         ) as ((stdout_fd, stderr_fd), writer),\
-         NailgunStreamStdinReader.open(sock, stdin_isatty) as stdin_fd,\
-         stdio_as(stdout_fd=stdout_fd, stderr_fd=stderr_fd, stdin_fd=stdin_fd):
-      # N.B. This will be passed to and called by the `DaemonExiter` prior to sending an
-      # exit chunk, to avoid any socket shutdown vs write races.
-      stdout, stderr = sys.stdout, sys.stderr
-      def finalizer():
-        try:
-          stdout.flush()
-          stderr.flush()
-        finally:
-          time.sleep(.001)  # HACK: Sleep 1ms in the main thread to free the GIL.
-          writer.stop()
-          writer.join()
-          stdout.close()
-          stderr.close()
-      yield finalizer
+    # If all stdio is a tty, there's only one logical I/O device (the tty device). This happens to
+    # be addressable as a file in OSX and Linux, so we take advantage of that and directly open the
+    # character device for output redirection - eliminating the need to directly marshall any
+    # interactive stdio back/forth across the socket and permitting full, correct tty control with
+    # no middle-man.
+    if all((stdin_isatty, stdout_isatty, stderr_isatty)):
+      stdin_ttyname, stdout_ttyname, stderr_ttyname = NailgunProtocol.ttynames_from_env(self._env)
+      assert stdin_ttyname == stdout_ttyname == stderr_ttyname, (
+        'expected all stdio ttys to be the same! '
```

Comment:
Include their names in the error to help us work out what's going on when they do file a bug?

Reply:
good call, fixed!